File: ProjectEulerPython/problems/problem_027.py
import math_helpers
from problems.problem import Problem

def f(n:int, a: int, b: int) -> int:
  """f(n) = n^2 + an + b"""
  return n * n + a * n + b

class Problem027(Problem):

  def calculate_answer(self) -> int:
    upper_bound_n = 100 # guess
    upper_bound_f = f(upper_bound_n, 99, 1000)
    prime_sieve = math_helpers.index_prime_by_sieve(upper_bound_f)

    #f(0) = b ==> b must be a (positive) prime
    primes_below_1000 = math_helpers.primes_below(1000)
    max_length = 0

    a = 1
    b = 1

    for b_inc in primes_below_1000:
      for a_inc in range(1, 1000, 2):
        for unit in range(-1, 3, 2):
          length = 0
          n = 0
          a_current = unit * a_inc
          while True:
            f_n = f(n, a_current, b_inc)
            if f_n < 0:
              f_n = -f_n
            if prime_sieve[f_n]:
              length += 1
            else:
              break
            n += 1
          if length > max_length:
            a = a_current
            b = b_inc
            max_length = length
            self.print_detail("length = " + str(length) + "; a = " + str(a) + "; b = " + str(b))

    return a * b

File: ProjectEuler/Problems/Problem027.cs
using Microsoft.Extensions.Logging;

namespace ProjectEuler.Problems
{
    public class Problem027 : Problem<long>
    {
        public Problem027(ILogger<Problem027> logger) : base(logger) { }

        protected override long CalculateAnswer()
        {
            var upper_bound_n = 100L; // guess
            var upper_bound_function = Function(upper_bound_n, 999L, 1000L);
            var composite_sieve = MathHelpers.IndexCompositeBySieve(upper_bound_function);

            //f(n) = n^2 + an + b ==> f(0) = b ==> b must be a (positive) prime
            var primes_below_1000 = MathHelpers.PrimesLessOrEqualTo(1000);
            var max_length = 0;

            var a = 1L;
            var b = 1L;
            foreach (var b_inc in primes_below_1000)
            {
                //since b is prime ==> b is odd ==> b + 1 is even; f(1) = 1 + a + b = a + (b + 1) which is even if a is even ==> a must be odd
                for (var a_inc = 1; a_inc < 1000; a_inc += 2)
                {
                    //this loop is to try both positive and negative values for a_loop
                    for (var unit = -1L; unit <= 1L; unit += 2L)
                    {
                        var length = 0;
                        var n = 0L;
                        var a_current = unit * a_inc;
                        while (true)
                        {
                            var f = Function(n, a_current, b_inc);
                            if (f < 0L) f = -f;
                            if (!composite_sieve[f]) ++length;
                            else break;
                            ++n;
                        }
                        if (length > max_length)
                        {
                            a = a_current;
                            b = b_inc;
                            max_length = length;
                            Logger.LogDebug("length = {length}; a = {a}; b = {b}", length, a, b);
                        }
                    }
                }
            }

            return a * b;
        }

        private static long Function(long n, long a, long b)
        {
            //f(n) = n^2 + an + b
            return n * n + a * n + b;
        }
    }
}
